### PR TITLE
[Datasets] Add spread resource prefix for manual round-robin resource-based task load balancing.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Adds a private spread resource prefix API for manual round-robin resource-based task load balancing, used for read and shuffle stages. Nodes (and therefore resources that match the spread prefix) that don't satisfy the task's resource requests are automatically filtered out, so a secondary custom resource or setting `num_cpus=0` on nodes can both be used to prevent these tasks from matching some subset of nodes (e.g. isolating a set of GPU workers from CPU workers).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/impl/util.py [hunk 1]
```diff
@@ -30,3 +35,66 @@ def _check_pyarrow_version():
                            "to ensure compatibility with Ray Datasets.")
         else:
             _VERSION_VALIDATED = True
+
+
+def _get_spread_resources_iter(nodes: List[Dict[str, Any]],
```
Comment (ericl) [inline python/ray/data/impl/util.py:40]: Would be nice to add unit test specifically for these helpers

File: python/ray/data/impl/shuffle.py [hunk 2]
```diff
@@ -11,40 +10,37 @@
...
-    # Create separate resource iterators for the map and reduce stages.
-    map_resource_iter, reduce_resource_iter = itertools.tee(
-        round_robin_resource_provider, 2)
+def simple_shuffle(
+        input_blocks: BlockList[T],
+        output_num_blocks: int,
+        *,
+        random_shuffle: bool = False,
+        random_seed: Optional[int] = None,
+        map_ray_remote_args: Optional[Dict[str, Any]] = None,
```
Comment (ericl) [inline python/ray/data/impl/shuffle.py:24]: Let's add a TODO that these hacks should be removed once we have proper feature support

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Merging to unblock, can address comments separately as optional

File: python/ray/data/impl/util.py [hunk 3]
```diff
@@ -30,3 +35,66 @@ def _check_pyarrow_version():
...
+                               spread_resource_prefix: str,
+                               ray_remote_args: Dict[str, Any]):
+    """Returns a round-robin iterator over resources that match the given
+    prefix and that coexist on nodes with the resource requests given in the
+    provided remote args (along with the task resource request defaults).
+    """
+    # Extract the resource labels from the remote args.
```
Comment (scv119) [inline python/ray/data/impl/util.py:47]: should we sort the nodes?

Comment (clarkzinzow) [reply python/ray/data/impl/util.py:47]: @scv119 The returned resource labels from `_filtered_resources()` are sorted!

**Pull request merged**